### ISSUE TITLE
kernel combination step that voldie missed

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#671]
https://gitter.im/grin_community/dev?at=5a71e52098927d57455df4a9

tl;dr If a valid tx consists of - 
* inputs
* output
* kernel(s)

And they sum to zero to be a valid tx.
Then it is possible to simply try all permutations of inputs and outputs for a given kernel to find a set that sum to zero (particularly given that most txs only have a few inputs and outputs).

From @tromp on gitter - 

> i just realized that making each block look like one huge tx doesn't help privacy much
the kernels still reveal the corresponding inputs and outputs

> you can still find inputs and outputs since they have to add up to 0 with the kernel?!

> i guess i'm having issues with this paragraph of the MW whitepaper:
Now, creating transactions in this manner supports OWAS already. To show this,
suppose we have two transactions that have a surplus k1G and k2G, and the
attached signatures with these. Then you can combine the lists of inputs and
outputs of the two transactions, with both k1G and k2G to the mix, and
voilá! is again a valid transaction. From the combination, it is impossible to
say which outputs or inputs are from which original transaction.

> it seems there's only one way to split inputs and outputs so that these form valid txs with their kernel

> most tx have very few ins and outs, so a polytime algorithm could recover them
e.g. suppose all tx have 2 in & 2 out. then just try all (n choose 4) subsets

Resolved by #681.

**Comment by antiochp** [comment]
One thing we discussed - 

Mitigate this via "kernel bloat" - build txs with many kernels. 
A large number of kernels would make the search space significantly larger, making recovery of individual txs harder.

Wallet could be configured to build txs with variable numbers of kernels based on level of anonymity desired.
Maybe you pay more in fees for more kernels?

**Comment by antiochp** [comment]
@apoelstra got paged about this and posted this link to earlier comment -
 https://www.reddit.com/r/Bitcoin/comments/4vub3y/mimblewimble_noninteractive_coinjoin_and_better/d61n41v/

**Comment by antiochp** [comment]
From @apoelstra in gitter - 

if you publish a kernel kG as (k1G, k2) where k1 + k2 = k, and a signature with k1G, that's as good as publishing kG and a signature w kG
as far as proving knowledge of the discrete log goes
but in the form (k1G, k2) you can partially combine kernels. like (k1G, k2) + (l1G, l2) becomes (k1G, l1G, (l2 + k2))

**Comment by antiochp** [comment]
cc @ignopeverell

**Comment by ignopeverell** [comment]
Right, I actually wanted to implement that fix way back when and completely forgot about it. Glad you 2 ran back into it!

**Comment by antiochp** [comment]
(still like our idea of generating 20 kernels per tx)

**Comment by ignopeverell** [comment]
What would multiple kernel give you assuming the offset is done? Btw if we do this, I was thinking we should also sum fees up, save a few bytes in kernels.

**Comment by antiochp** [comment]
Nothing... but it would have made it "harder" to identify inputs/outputs if we couldn't offset the kernels. At the cost of thousands and thousands of extra kernels...

**Comment by antiochp** [comment]
Agreed on summing the fees up as well.
Right now the fees leak _some_ info about the txs in the block (relative numbers of inputs/outputs, complexity of tx).

**Comment by antiochp** [comment]
Digging through the secp code and it looks like we can _add_ two secret keys together -

https://github.com/mimblewimble/secp256k1-zkp/blob/960034ca957ac630dcefdf5a285c92ca830ea7ab/src/eckey_impl.h#L55

Which calls the low level - 

https://github.com/mimblewimble/secp256k1-zkp/blob/960034ca957ac630dcefdf5a285c92ca830ea7ab/src/scalar_low_impl.h#L35


Does splitting a key (so we can do `k = k1 + k2`) involve creating a new secret key `k2` and subtracting it from `k` to give `k1 = k - k2`?

Can we subtract secret keys?

I see we can do what appears to be low level subtraction of `secp256k1_scalar` values - 

https://github.com/mimblewimble/secp256k1-zkp/blob/960034ca957ac630dcefdf5a285c92ca830ea7ab/src/scalar_low_impl.h#L40

**Comment by antiochp** [comment]
#681 for some initial work around splitting the key (work in progress).
